File: scala/CardGame/src/test/scala/es/juanc/katas/cardgame/CardGameSpec.scala
package es.juanc.katas.cardgame

import es.juanc.katas.UnitSpec

/**
  * TODO list:
  * - entorno configurado para ejecutar las especificaciones
  * - dado 1 carta to cada jugador gana el jugador uno: "Win player one: 1 to 0"
  * - dada 1 carta to cada jugador empatan: "Tie"
  * - dada 1 carta to cada jugador gana el jugador dos: "Win player two: 1 to 0"
  * - dada 2 cartas to cada jugador gana el jugador dos: "Win player two: 2 to 0"
  * - dada 3 cartas to cada jugador gana el jugador dos: "Win player two: 2 to 1"
  * - ...
  */

class CardGameSpec extends UnitSpec {

  "Spec" must "pass test" in {

    "" shouldBe empty
  }

  "dado N carta to cada jugador" must "se obtiene el resultado parametrizado, con resultado en ingles" in {

    val values =
      Table(
        ("deck1",                         "deck2",                          "expected"),
        (List("12"),                      List("1"),                        "Win player one: 1 to 0"),
        (List("1"),                       List("12"),                       "Win player two: 1 to 0"),
        (List("12", "11"),                List("2", "1"),                   "Win player one: 2 to 0"),
        (List("1", "2"),                  List("12", "11"),                 "Win player two: 2 to 0"),
        (List("1", "12", "2"),            List("12", "11", "10"),           "Win player two: 2 to 1"),
        (List("12", "11", "10"),          List("1", "12", "2"),             "Win player one: 2 to 1"),
        (List("12", "11", "10", "2"),     List("1", "12", "2", "10"),       "Tie"),
        (List("11", "1", "10", "2", "1"), List("11", "1", "10", "10", "5"), "Win player two: 5 to 3")
      )

    forAll(values) { (deck1: List[String], deck2: List[String], expected: String) =>

      CardGameInEn.whoWins(deck1, deck2) shouldBe expected
    }
  }

  "dado 1 carta to cada jugador" must "gana el jugador uno, con resultado en español" in {

    CardGameInEs.whoWins(List("12"), List("1")) should equal("Gana el jugador uno: 1 a 0")
  }
}

File: javascript/CardGame/test/CardGameSpec.js
/**
 * TODO list:
 * - entorno configurado para ejecutar las especificaciones
 * - dado 1 carta a cada jugador gana el jugador uno: "Win player one: 1 a 0"
 * - dada 1 carta a cada jugador empatan: "Tie"
 * - dada 1 carta a cada jugador gana el jugador dos: "Win player two: 1 a 0"
 * - dada 2 cartas a cada jugador gana el jugador dos: "Win player two: 2 a 0"
 * - dada 3 cartas a cada jugador gana el jugador dos: "Win player two: 2 a 1"
 * - ...
 */

const chai = require('chai')
const expect = chai.expect

const CardGame = require('../src/CardGame')

describe('Configuracion del entorno', function() {
  it('deberia pasar el test cuando el entorno este correctamente configurado', function() {
    expect(true).to.be.eq(true)
  })
})

describe("Especificaciones del 'CardGame'", function() {

  const spectation = ({description, given, then}) => {

    const whoWins = CardGame().whoWins(given.deckPlayer1, given.deckPlayer2)

    it(description, function() {
      expect(whoWins).to.equal(then)
    })
  }

  describe('cuando se reparte una carta', function() {
    [
      {
        description: 'mayor al jugador uno deberia ganar el jugador uno: 1 a 0',
        given: {
          deckPlayer1: ['K'],
          deckPlayer2: ['1'],
        },
        then: 'Win player one: 1 a 0',
      },
      {
        description: 'mayor al jugador dos deberia ganar el jugador dos: 1 a 0',
        given: {
          deckPlayer1: ['1'],
          deckPlayer2: ['K'],
        },
        then: 'Win player two: 1 a 0',
      },
      {
        description: 'iguales a ambos jugadores empatan',
        given: {
          deckPlayer1: ['K'],
          deckPlayer2: ['K'],
        },
        then: 'Tie',
      },
    ].forEach(spectation)
  })

  describe('cuando se reparte 2 cartas', function() {
    [
      {
        description: 'ambas mayores al jugador dos este gana: 2 a 0',
        given: {
          deckPlayer1: ['K', 'Q'],
          deckPlayer2: ['1', '2'],
        },
        then: 'Win player one: 2 a 0',
      },
      {
        description: 'iguales a ambos jugadores empatan',
        given: {
          deckPlayer1: ['K', '1'],
          deckPlayer2: ['K', '1'],
        },
        then: 'Tie',
      },
    ].forEach(spectation)
  })

  describe('cuando se reparte 3 cartas', function() {
    [
      {
        description: 'todas mayores al jugador uno este gana: 3 a 0',
        given: {
          deckPlayer1: ['K', 'Q', 'J'],
          deckPlayer2: ['1', '2', '3'],
        },
        then: 'Win player one: 3 a 0',
      },
      {
        description: 'gana el jugador dos: 2 a 1',
        given: {
          deckPlayer1: ['2', '1', '7'],
          deckPlayer2: ['K', 'Q', '1'],
        },
        then: 'Win player two: 2 a 1',
      },
      {
        description: 'gana el jugador uno: 2 a 1',
        given: {
          deckPlayer1: ['K', '1', 'Q'],
          deckPlayer2: ['1', 'K', '2'],
        },
        then: 'Win player one: 2 a 1',
      },
    ].forEach(spectation)
  })
})
